depend/pari/doc/gp.1
.TH GP 1 "11 September 2017"
.SH NAME
gp \- The PARI calculator
.SH SYNOPSIS
.B gp
.RB [ -s
.IR stacksize ]
.RB [ -p
.IR primelimit ]
.RB [ --emacs ]
.RB [ -f | --fast ]
.RB [ -q | --quiet ]
.RB [ -D | --default
.IR key=val ]
.RB [ --help ]
.RB [ --test ]
.RB [ --texmacs ]
.RB [ --version ]
.RB [ --version-short ]
[ file1 file2 ...]

.SH DESCRIPTION
Invokes the PARI-GP calculator
\&\fBgp\fR; gp is an advanced programmable calculator, specializing in number
theory, which computes symbolically as long as possible, numerically where
needed, and contains a wealth of arithmetic functions: factorizations,
elliptic curves, Galois theory, class field theory, modular forms, etc.
Commands, written in the GP scripting language, are input interactively or
loaded from files.

If present at the end of the command line, files 'file1', 'file2', ...
are loaded on startup; they must be written in the GP language.

.SH OPTIONS
Command line options are available in both short form (-f) and POSIX-like
(--fast). Numeric arguments can be followed by a modifier
.B k
,
.B M
or
.B G
at the user's convenience; in that case the argument is multiplied by 10^3,
10^6, or 10^9 respectively.

.TP
.B \-f, \--fast
Fast start (or factory settings). Do not read
.B .gprc
(see below) upon startup.
.TP
.B \-p limit
[DEPRECATED]
Upon startup, gp computes a table of small primes used in
number-theoretic applications. If
.I primelimit
is set, the table include primes up to that bound instead of the default
(= 500000). It is now mostly useless to change this value.
.TP
.B \-q, \--quiet
Quiet mode. Do not print headers or history numbers and do not say goodbye.

.TP
.B \-D, \--default key=val
performs
.BR default(key,
.BR val) ";"
on startup, overriding values from the
.B gprc
preferences file. 'val' must be a constant value and is not allowed to
involve any computation (e.g. 1+1 is forbidden). Any number of such
default-setting statements may appear on the command line. A key may be set
multiple times, the last setting taking precedence

.TP
.B \-s limit
Size of gp internal stack allocated on startup. When gp runs out of space, it
interrupts the current computation and raises a
.BI "stack overflow"
exception. If this occurs frequently, start with a bigger stack. The stack
size can also be
increased from within gp, using
.BR default(parisize, limit) ";"
it may be convenient to set
.B stacksize
from your
.B .gprc.
Note that computations with a
.B smaller
stack may be more efficient due to better data locality. Most computations
should need less than 20MB.

.TP
.B \--emacs
gp can be run in an
.I Emacs
shell (see GP User's manual for details). This flag is then required for
smooth interaction with the
.I PariEmacs
package (pari.el). It is set automatically by the pari.el package, and will
produce display oddities if you set it outside of an
.I Emacs
session.
.TP
.B \--help
print a summary of available command-line options.
.TP
.B \--test
run gp in test mode: suppress printing of history numbers and wrap long
output lines (to get readable diff output). For benches only.
.TP
.B \--texmacs
gp can be run from a
.I TeXmacs
frontend. This flag is set by TeXmacs, to enable special purpose
communication channels. Do not set it yourself.

.TP
.B \--version
output version info (banner) then exit.

.TP
.B \--version-short
output version number then exit.

.SH USE
.TP
.B ?
to get online help.
.TP
.B ??
to get extended online help (more precisely, to call the external help
program,
.B gphelp
by default)
.TP
.B quit
(or \\q), or
.B EOF
(Ctrl-D) to quit
.BR gp .
.PP
The following works only when gp was linked with GNU
.IR readline
library:
.TP
arrow keys
for editing and viewing the input history.
.TP
.B TAB
 for automatic completion

.SH MANUALS
The following material is included in the standard distribution (originally
in TeX format) and can also be downloaded at
.RS
.I http://pari.math.u-bordeaux.fr/doc.html
.RE

.TP
.I The User's Guide to PARI/GP
(users.dvi)
.TP
.I The User's Guide to the PARI library
(library.dvi)
.TP
.I The Developer's Guide to the PARI library
(develop.dvi)
.TP
.I PARI/GP, a tutorial
(tutorial.dvi)
.TP
.I PARI/GP reference cards
(refcard*.dvi)

.SH FILES
.TP
.I gp
main executable
.TP
.I $HOME/.gprc
(or $GPRC if set) user preference file, read at beginning of execution by
each
.B gp
shell. A default gprc
.I gprc.dft
is provided with the distribution. If this file cannot be found,
.I /etc/gprc
is checked instead.

.TP
.I <logfile>
a file used to log in all commands and results; default:
.B pari.log
(you need to set the
.B log
default in your gprc or interactively)

.TP
.I <psfile>
a file used to dump PostScript drawings; default:
.B pari.ps

.TP
.I <histfile>
a file where gp will keep a history of all input commands (you need to set
the
.B histfile
default in the gprc file)

.TP
.I gphelp
default external help program (as above)
.TP
.I *.gp
GP programs

.SH ENVIRONMENT
.TP
.I $GPRC
place to look for the user's preference file (gprc); if the file does not exist,
we then check in $HOME/.gprc, /etc/gprc, and finally for a file named 'gprc'
in PARI's
.B datadir.

.TP
.I $GP_DATA_DIR
directory containing data installed by optional PARI packages.
For example, the Galois resolvents files in directory
.I galdata/
needed by the
.B polgalois
function, in degrees 8 to 11; or the modular polynomials in
.I seadata/
used by the
.B ellap
function for large base fields. This environment variable
overrides PARI's 'datadir', defined at Configure time.

.TP
.I $GP_POSTSCRIPT_VIEWER
an application able to display PostScript files, used by the
.I plotps
graphic engine. This engine is a fallback used to output hi-res plots even
when no compatible graphical library was available on your platform at
Configure time. (Dumps the graph to a temporary file, then open the file.)

.TP
.I $GP_SVG_VIEWER
an application able to display SGV images files, used by the
.I plotsvg
graphic engine. This engine is a fallback used to output hi-res plots even
when no compatible graphical library was available on your platform at
Configure time. (Dumps the graph to a temporary file, then open the file.)

.TP
.I $GPHELP
name of the external help program invoked by ?? and ??? shortcuts.

.TP
.I $GPTMPDIR
name of the directory where temporary files will be generated.

.SH HOME PAGE
PARI's home page resides at
.RS
.I http://pari.math.u-bordeaux.fr/
.RE

.SH MAILING LISTS
There are a number of mailing lists devoted to the PARI/GP package, and most
feedback should be directed to those. See
.RS
.I http://pari.math.u-bordeaux.fr/lists-index.html
.RE
for details. The most important ones are:

.PP
-
.B pari-announce
(moderated): for us to announce major version changes.
.PP
-
.B pari-dev:
for everything related to the development of PARI, including
suggestions, technical questions, bug reports or patch submissions.

.PP
-
.B pari-users:
for discuss about everything else, in particular ask for help.

To subscribe, send empty messages with a Subject: containing the word
"subscribe" respectively to

.PP
   pari-announce-request@pari.math.u-bordeaux.fr
.PP
   pari-users-request@pari.math.u-bordeaux.fr
.PP
   pari-dev-request@pari.math.u-bordeaux.fr

.SH BUG REPORTS
Bugs should be submitted online to our Bug Tracking System, available from
PARI's home page, or directly from the URL
.RS
.I http://pari.math.u-bordeaux.fr/Bugs/
.RE
Further instructions can be found on that page.

.SH TRIVIA
Despite the leading G, GP has nothing to do with GNU. The first version was
originally called GPC, for Great Programmable Calculator. For some reason,
the trailing C was eventually dropped.

PARI has nothing to do with the French capital. The name is a pun about the
project's early stages when the authors started to implement a library for
"Pascal ARIthmetic" in the PASCAL programming language. They quickly
switched to C.

For the benefit of non-native French speakers, here's a slightly expanded
explanation:
.B Blaise Pascal
(1623-1662) was a famous French mathematician and philosopher who was one
of the founders of probability and devised one of the first "arithmetic
machines". He once proposed the following "proof" of the existence of God
for the unbelievers: whether He exists or not I lose nothing by believing
in Him, whereas if He does and I misbehave... This is the so-called "pari
de Pascal" (Pascal's Wager).

Note that PARI also means "fairy" in Persian.

.SH AUTHORS
PARI was originally written by Christian Batut, Dominique Bernardi, Henri
Cohen, and Michel Olivier in Laboratoire A2X (Universite Bordeaux I, France),
and was maintained by Henri Cohen up to version 1.39.15 (1995), and by Karim
Belabas since then.

A great number of people have contributed to the successive improvements
which eventually resulted in the present version. See the AUTHORS file in
the distribution.

.SH SEE ALSO
.IR gap (1),
.IR gphelp (1),
.IR perl (1),
.IR readline (3),
.IR sage (1),
.IR tex (1),
.IR texmacs (1),

.SH COPYING

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 675 Mass
Ave, Cambridge, MA 02139, USA.


depend/pari-2.11.2/src/gp/gp.c
/* Copyright (C) 2000  The PARI group.

This file is part of the PARI/GP package.

PARI/GP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA. */

/*******************************************************************/
/**                                                               **/
/**                        PARI CALCULATOR                        **/
/**                                                               **/
/*******************************************************************/
#ifdef _WIN32
#  include "../systems/mingw/pwinver.h"
#  include <windows.h>
#  include "../systems/mingw/mingw.h"
#endif
#include "pari.h"
#include "paripriv.h"
#include "gp.h"

static jmp_buf *env;
static pari_stack s_env;
void (*cb_gp_output)(GEN z) = NULL;
void (*cb_pari_end_output)(void) = NULL;

static void
gp_ask_confirm(const char *s)
{
  err_printf(s);
  err_printf(". OK ? (^C if not)\n");
  pari_hit_return();
}

/* numerr < 0: from SIGINT */
static void
gp_err_recover(long numerr)
{
  longjmp(env[s_env.n-1], numerr);
}
/* numerr < 0: from SIGINT */
static void
gp_pre_recover(long numerr)
{
  if (numerr>=0)
  {
    out_puts(pariErr, "\n"); pariErr->flush();
  }
  longjmp(env[s_env.n-1], numerr);
}

static void
reset_ctrlc(void)
{
#if defined(_WIN32) || defined(__CYGWIN32__)
  win32ctrlc = 0;
#endif
}

static int
is_silent(char *s) { return s[strlen(s) - 1] == ';'; }

static int stdin_isatty = 0;
static int
is_interactive(void)
{ return pari_infile == stdin && stdin_isatty; }

/*******************************************************************/
/**                                                               **/
/**                        INITIALIZATION                         **/
/**                                                               **/
/*******************************************************************/
static void
print_shortversion(void)
{
  const ulong mask = (1UL<<PARI_VERSION_SHIFT) - 1;
  ulong n = paricfg_version_code, major, minor, patch;

  patch = n & mask; n >>= PARI_VERSION_SHIFT;
  minor = n & mask; n >>= PARI_VERSION_SHIFT;
  major = n;
  printf("%lu.%lu.%lu\n", major,minor,patch); exit(0);
}

static void
usage(char *s)
{
  printf("### Usage: %s [options] [GP files]\n", s);
  printf("Available options:\n");
  printf("  [-f,--fast]\t\tFast start: do not read .gprc\n");
  printf("  [-q,--quiet]\t\tQuiet mode: do not print banner and history numbers\n");
  printf("  [-s stacksize]\tStart with the PARI stack of given size (in bytes)\n");
  printf("  [--default key=val]\tExecute default(key,val) on startup\n");
  printf("  [--emacs]\t\tRun as if in Emacs shell\n");
  printf("  [--help]\t\tPrint this message\n");
  printf("  [--test]\t\tTest mode. No history, wrap long lines (bench only)\n");
  printf("  [--texmacs]\t\tRun as if using TeXmacs frontend\n");
  printf("  [--version]\t\tOutput version info and exit\n");
  printf("  [--version-short]\tOutput version number and exit\n\n");
  exit(0);
}

static void
gp_head(void)
{
  pari_print_version();
  pari_putc('\n');
  pari_center("Copyright (C) 2000-2018 The PARI Group");
  pari_putc('\n');
  print_text("PARI/GP is free software, covered by the GNU General Public \
License, and comes WITHOUT ANY WARRANTY WHATSOEVER.");
  pari_puts("\nType ? for help, \\q to quit.\n");
  pari_printf("Type ?%d for how to get moral"
              " (and possibly technical) support.\n", pari_community());
  if (pari_mainstack->vsize)
    pari_printf("\nparisizemax = %lu, primelimit = %lu",
                pari_mainstack->vsize,GP_DATA->primelimit);
  else
    pari_printf("\nparisize = %lu, primelimit = %lu",
                pari_mainstack->rsize,GP_DATA->primelimit);
  if (pari_mt_nbthreads > 1)
    pari_printf(", nbthreads = %lu", pari_mt_nbthreads);
  pari_putc('\n');
}

static char *
read_arg(long *nread, char *t, long argc, char **argv)
{
  long i = *nread;
  if (isdigit((int)*t)) return t;
  if (*t || i==argc) usage(argv[0]);
  *nread = i+1; return argv[i];
}

static char *
read_arg_equal(long *nread, char *t, long argc, char **argv)
{
  long i = *nread;
  if (*t=='=' && isdigit((int)t[1])) return t+1;
  if (*t || i==argc) usage(argv[0]);
  *nread = i+1; return argv[i];
}

static void
init_trivial_stack(void)
{
  const size_t s = 2048;
  pari_mainstack->size = s;
  pari_mainstack->bot = (pari_sp)pari_malloc(s);
  avma = pari_mainstack->top = pari_mainstack->bot + s;
}

static void
free_trivial_stack(void)
{
  free((void*)pari_mainstack->bot);
}

typedef struct { char *key, *val; } pair_t;
/* If ab of the form key=val, record pair in new stack entry
 * P[n].key must be freed by caller to avoid memory leak */
static void
record_default(pari_stack *s_P, char *ab)
{
  pair_t *P = (pair_t*)*pari_stack_base(s_P);
  char *k, *v;
  long n;
  ab = pari_strdup(ab);
  parse_key_val(ab, &k, &v);
  n = pari_stack_new(s_P);
  P[n].key = k;
  P[n].val = v;
}
static void
read_opt(pari_stack *p_A, long argc, char **argv)
{
  pair_t *P;
  pari_stack s_P; /* key / value to record default() settings */
  char *b = NULL, *p = NULL, *s = NULL;
  ulong f = GP_DATA->flags;
  long i = 1, initrc = 1;

  (void)&p; (void)&b; (void)&s; /* -Wall gcc-2.95 */

  pari_stack_init(&s_P,sizeof(*P),(void**)&P);
  pari_stack_alloc(&s_P, 64);
  pari_outfile = stderr;
  while (i < argc)
  {
    char *t = argv[i];

    if (*t++ != '-') break;
    i++;
START:
    switch(*t++)
    {
      case 'p': p = read_arg(&i,t,argc,argv); break;
      case 's': s = read_arg(&i,t,argc,argv); break;
      case 'e':
        f |= gpd_EMACS; if (*t) goto START;
        break;
      case 'q':
        f |= gpd_QUIET; if (*t) goto START;
        break;
      case 't':
        f |= gpd_TEST; if (*t) goto START;
        break;
      case 'f':
        initrc = 0; if (*t) goto START;
        break;
      case 'D':
        if (*t || i == argc) usage(argv[0]);
        record_default(&s_P, argv[i++]);
        break;
      case '-':
        if (strcmp(t, "version-short") == 0) { print_shortversion(); exit(0); }
        if (strcmp(t, "version") == 0) {
          init_trivial_stack(); pari_print_version();
          free_trivial_stack(); exit(0);
        }
        if (strcmp(t, "default") == 0) {
          if (i == argc) usage(argv[0]);
          record_default(&s_P, argv[i++]);
          break;
        }
        if (strcmp(t, "texmacs") == 0) { f |= gpd_TEXMACS; break; }
        if (strcmp(t, "emacs") == 0) { f |= gpd_EMACS; break; }
        if (strcmp(t, "test") == 0) { f |= gpd_TEST; initrc = 0; break; }
        if (strcmp(t, "quiet") == 0) { f |= gpd_QUIET; break; }
        if (strcmp(t, "fast") == 0) { initrc = 0; break; }
        if (strncmp(t, "primelimit",10) == 0) {p = read_arg_equal(&i,t+10,argc,argv); break; }
        if (strncmp(t, "stacksize",9) == 0) {s = read_arg_equal(&i,t+9,argc,argv); break; }
       /* fall through */
      default:
        usage(argv[0]);
    }
  }
  if (f & gpd_TEST) stdin_isatty = 0;
  GP_DATA->flags = f;
#ifdef READLINE
  GP_DATA->use_readline = stdin_isatty;
#endif
  if (!is_interactive()) GP_DATA->breakloop = 0;
  if (initrc) gp_initrc(p_A);
  for ( ; i < argc; i++) pari_stack_pushp(p_A, pari_strdup(argv[i]));

  /* override the values from gprc */
  if (p) (void)sd_primelimit(p, d_INITRC);
  if (s) (void)sd_parisize(s, d_INITRC);
  for (i = 0; i < s_P.n; i++) {
    setdefault(P[i].key, P[i].val, d_INITRC);
    free((void*)P[i].key);
  }
  pari_stack_delete(&s_P);
  pari_outfile = stdout;
}

/*******************************************************************/
/**                                                               **/
/**                            TEST MODE                          **/
/**                                                               **/
/*******************************************************************/
static int
test_is_interactive(void) { return 0; }

static void
test_output(GEN z) { init_linewrap(76); gen_output(z); }
void
init_test(void)
{
  disable_color = 1;
  init_linewrap(76);
  pari_errfile = stdout;
  cb_gp_output = test_output;
  cb_pari_is_interactive = test_is_interactive;
}

/*******************************************************************/
/**                                                               **/
/**                   FORMAT GP OUTPUT                            **/
/**                                                               **/
/*******************************************************************/
    /* REGULAR */
static void
normal_output(GEN z, long n)
{
  long l = 0;
  char *s;
  /* history number */
  if (n)
  {
    char buf[64];
    if (!(GP_DATA->flags & gpd_QUIET))
    {
      term_color(c_HIST);
      sprintf(buf, "%%%ld = ", n);
      pari_puts(buf);
      l = strlen(buf);
    }
  }
  /* output */
  term_color(c_OUTPUT);
  s = GENtostr(z);
  if (GP_DATA->lim_lines)
    lim_lines_output(s, l, GP_DATA->lim_lines);
  else
    pari_puts(s);
  pari_free(s);
  term_color(c_NONE); pari_putc('\n');
}

static void
gp_output(GEN z)
{
  if (cb_gp_output) { cb_gp_output(z); return; }
  if (GP_DATA->fmt->prettyp == f_PRETTY)
  { if (tex2mail_output(z, GP_DATA->hist->total)) return; }
  normal_output(z, GP_DATA->hist->total);
  pari_flush();
}

static GEN
gp_main_loop(long ismain)
{
  VOLATILE GEN z = gnil;
  VOLATILE long t = 0;
  VOLATILE pari_sp av = avma;
  filtre_t F;
  Buffer *b = filtered_buffer(&F);
  struct gp_context rec;
  long er;
  if ((er = setjmp(env[s_env.n-1])))
  { /* recover: jump from error [ > 0 ] or allocatemem [ -1 ] */
    if (er > 0) { /* true error */
      if (!(GP_DATA->recover)) exit(1);
      gp_context_restore(&rec);
      /* true error not from main instance, let caller sort it out */
      if (!ismain) { kill_buffers_upto_including(b); return NULL; }
    } else { /* allocatemem */
      tmp_restore(rec.file.file);
      gp_context_save(&rec);
    }
    avma = av = pari_mainstack->top;
    parivstack_reset();
    kill_buffers_upto(b);
    pari_alarm(0);
  }
  for(;;)
  {
    gp_context_save(&rec);
    if (! gp_read_line(&F, NULL))
    {
      if (popinfile()) gp_quit(0);
      if (ismain) continue;
      pop_buffer(); return z;
    }
    if (ismain)
    {
      reset_ctrlc();
      timer_start(GP_DATA->T);
      pari_set_last_newline(1);
    }
    if (gp_meta(b->buf,ismain)) continue;
    z = pari_compile_str(b->buf);
    z = closure_evalres(z);
    if (!ismain) continue;

    t = timer_delay(GP_DATA->T);
    if (!pari_last_was_newline()) pari_putc('\n');
    pari_alarm(0);
    if (t && GP_DATA->chrono)
    {
      pari_puts("time = ");
      pari_puts(gp_format_time(t));
    }
    if (GP_DATA->simplify) z = simplify_shallow(z);
    pari_add_hist(z, t);
    if (z != gnil && ! is_silent(b->buf) ) gp_output(z);
    avma = av;
    parivstack_reset();
  }
}

/* as gp_read_file, before running the main gp instance */
static void
read_main(const char *s)
{
  GEN z;
  if (setjmp(env[s_env.n-1]))
    z = NULL;
  else {
    FILE *f = switchin(s);
    if (file_is_binary(f)) {
      z = readbin(s,f, NULL);
      popinfile();
    }
    else z = gp_main_loop(0);
  }
  if (!z) err_printf("... skipping file '%s'\n", s);
  avma = pari_mainstack->top;
}

static const char *
break_loop_prompt(long n)
{
  const char *s = (n==1)? "break> ": stack_sprintf("break[%ld]> ", n);
  return gp_format_prompt(s);
}

static long frame_level=0, dbg_level = 0;

static int
break_loop(int numerr)
{
  filtre_t F;
  Buffer *b;
  int sigint = numerr<0, go_on = sigint;
  struct gp_context rec1, rec2;
  const char *prompt, *msg;
  long nenv, oldframe_level = frame_level;
  pari_sp av;

  if (numerr == e_SYNTAX) return 0;
  if (numerr == e_STACK) { evalstate_clone(); avma = pari_mainstack->top; }
  gp_context_save(&rec1);

  b = filtered_buffer(&F);
  nenv=pari_stack_new(&s_env);
  prompt = break_loop_prompt(s_env.n-1);
  iferr_env = NULL;
  dbg_level = 0;
  frame_level = closure_context(oldframe_level, dbg_level);
  pari_infile = newfile(stdin, "stdin", mf_IN)->file;
  term_color(c_ERR); pari_putc('\n');
  if (sigint)
    msg = "Break loop: <Return> to continue; 'break' to go back to GP prompt";
  else
    msg = "Break loop: type 'break' to go back to GP prompt";
  print_errcontext(pariOut, msg, NULL, NULL);
  term_color(c_NONE);
  av = avma;
  for(;;)
  {
    GEN x;
    long er, br_status;
    avma = av;
    gp_context_save(&rec2);
    if ((er=setjmp(env[nenv])))
    {
      if (er < 0)
      {
        s_env.n = 1;
        frame_level = oldframe_level;
        longjmp(env[s_env.n-1], er);
      }
      gp_context_restore(&rec2);
      iferr_env = NULL;
      closure_err(dbg_level);
      compilestate_restore(&rec1.eval.comp);
      (void) closure_context(oldframe_level, dbg_level);
      pari_infile = newfile(stdin, "stdin", mf_IN)->file;
    }
    term_color(c_NONE);
    if (!gp_read_line(&F, prompt))
      br_status = br_BREAK; /* EOF */
    else
    {
      /* Empty input ? Continue if entry on sigint (exit debugger frame) */
      if (! *(b->buf) && sigint) break;
      reset_ctrlc();
      if (gp_meta(b->buf,0)) continue;
      x = pari_compile_str(b->buf);
      x = closure_evalbrk(x, &br_status);
    }
    switch (br_status)
    {
      case br_NEXT: case br_MULTINEXT:
        popinfile(); /* exit frame. Don't exit debugger if s_env.n > 2 */
        go_on = 0; goto BR_EXIT;
      case br_BREAK: case br_RETURN:
        killallfiles(); /* completely exit the debugger */
        go_on = 0; goto BR_EXIT;
    }
    if (x!=gnil && !is_silent(b->buf)) { term_color(c_OUTPUT); gen_output(x); }
  }
BR_EXIT:
  s_env.n=nenv;
  frame_level = oldframe_level;
  gp_context_restore(&rec1);
  pop_buffer(); return go_on;
}

#ifdef __CYGWIN32__
void
cyg_environment(int argc, char ** argv)
{
  char *ti_dirs = getenv("TERMINFO_DIRS");
  char *argv0, *p;
  char *newdir;
  long n;

  if (!argc || !argv) return;
  argv0 = *argv;
  if (!argv0 || !*argv0) return;
  p = strrchr(argv0, '/');
  if (!p)
    p = argv0 = "";
  else
    p++;
  n = p - argv0;
  if (ti_dirs)
  {
    n += 14 + strlen(ti_dirs) + 1 + 8 + 1;
    newdir = malloc(n);
    if (!newdir) return;
    snprintf(newdir, n-8, "TERMINFO_DIRS=%s:%s", ti_dirs, argv0);
  }
  else
  {
    n += 14 + 8 + 1;
    newdir = malloc(n);
    if (!newdir) return;
    snprintf(newdir, n-8, "TERMINFO_DIRS=%s", argv0);
  }
  strcpy(newdir+n-9,"terminfo");
  putenv(newdir);
}
#endif

int
main(int argc, char **argv)
{
  char **A;
  pari_stack s_A;

  GP_DATA = default_gp_data();
  pari_stack_init(&s_env, sizeof(*env), (void**)&env);
  (void)pari_stack_new(&s_env);

  if (setjmp(env[s_env.n-1]))
  {
    puts("### Errors on startup, exiting...\n\n");
    exit(1);
  }
#ifdef __CYGWIN32__
  cyg_environment(argc, argv);
#endif
  stdin_isatty = pari_stdin_isatty();
  pari_init_defaults();
  pari_library_path = DL_DFLT_NAME;
  pari_stack_init(&s_A,sizeof(*A),(void**)&A);
  pari_init_opts(1000000 * sizeof(long), 0, INIT_SIGm | INIT_noPRIMEm | INIT_noIMTm);
  cb_pari_err_recover = gp_err_recover;
  cb_pari_pre_recover = gp_pre_recover;
  cb_pari_break_loop = break_loop;
  cb_pari_is_interactive = is_interactive;

  read_opt(&s_A, argc,argv);
  pari_init_primes(GP_DATA->primelimit);
#ifdef SIGALRM
  (void)os_signal(SIGALRM,gp_alarm_handler);
#endif
  pari_add_module(functions_gp);

  pari_set_plot_engine(gp_get_plot);
  cb_pari_quit = gp_quit;
  cb_pari_whatnow = whatnow;
  cb_pari_sigint = gp_sigint_fun;
  cb_pari_handle_exception = gp_handle_exception;
  cb_pari_ask_confirm = gp_ask_confirm;
  pari_init_paths();
  pari_mt_init(); /* MPI: will not return on slaves (pari_MPI_rank = 0) */
#ifdef _WIN32
  if (stdin_isatty) win32_set_codepage();
#endif
#ifdef READLINE
  init_readline();
#endif
  if (GP_DATA->flags & gpd_EMACS) init_emacs();
  if (GP_DATA->flags & gpd_TEXMACS) init_texmacs();

  timer_start(GP_DATA->T);
  if (!(GP_DATA->flags & gpd_QUIET)) gp_head();
  if (GP_DATA->flags & gpd_TEST) init_test();
  if (s_A.n)
  {
    FILE *l = pari_logfile;
    long i;
    pari_logfile = NULL;
    for (i = 0; i < s_A.n; pari_free(A[i]),i++) read_main(A[i]);
    /* Reading one of the input files above can set pari_logfile.
     * Don't restore in that case. */
    if (!pari_logfile) pari_logfile = l;
  }
  pari_stack_delete(&s_A);
  (void)gp_main_loop(1);
  gp_quit(0);
  return 0; /* LCOV_EXCL_LINE */
}

void
pari_breakpoint(void)
{
  if (!pari_last_was_newline()) pari_putc('\n');
  closure_err(0);
  if (cb_pari_break_loop && cb_pari_break_loop(-1)) return;
  cb_pari_err_recover(e_MISC);
}

void
dbg_down(long k)
{
  if (k<0) k=0;
  dbg_level -= k;
  if (dbg_level<0) dbg_level=0;
  gp_err_recover(e_NONE);
}

GEN
dbg_err(void) { GEN E = pari_err_last(); return E? gcopy(E):gnil; }

void
dbg_up(long k)
{
  if (k<0) k=0;
  dbg_level += k;
  if (dbg_level>frame_level) dbg_level=frame_level;
  gp_err_recover(e_NONE);
}

void
gp_quit(long code)
{
  pari_kill_plot_engine();
  pari_close();
  kill_buffers_upto(NULL);
  if (!(GP_DATA->flags & gpd_QUIET)) pari_puts("Goodbye!\n");
  if (cb_pari_end_output) cb_pari_end_output();
  exit(code);
}

void
whatnow0(char *s) { whatnow(pariOut, s,0); }

#include "gp_init.h"
